# Fingerprints

In [5]:
from rdkit.Chem import AllChem as Chem
from rdkit.Chem import Draw
from rdkit.Chem.Draw import IPythonConsole
import csv

In [6]:
with open('Epidermal_growth_factor_receptor_erbB1_IC50.csv', 'r') as csvfile:
    reader = csv.DictReader(csvfile, delimiter=";")
    EGFR_ligands = [Chem.MolFromSmiles(m['Smiles']) for m in reader]

suppl = Chem.SDMolSupplier('../data/drugbank.sdf')
drugs = [m for m in suppl if m]

RDKit ERROR: [17:25:00] ERROR: Explicit valence for atom # 0 C greater than permitted
RDKit ERROR: [17:58:07] Explicit valence for atom # 7 N, 4, is greater than permitted
RDKit ERROR: [17:58:07] ERROR: Could not sanitize molecule ending on line 20009
RDKit ERROR: [17:58:07] ERROR: Explicit valence for atom # 7 N, 4, is greater than permitted
RDKit ERROR: [17:58:07] Can't kekulize mol.  Unkekulized atoms: 1 2 3 5 6 7 8 9 10
RDKit ERROR: 
RDKit ERROR: [17:58:07] ERROR: Could not sanitize molecule ending on line 250947
RDKit ERROR: [17:58:07] ERROR: Can't kekulize mol.  Unkekulized atoms: 1 2 3 5 6 7 8 9 10
RDKit ERROR: 
RDKit ERROR: [17:58:07] Explicit valence for atom # 17 O, 3, is greater than permitted
RDKit ERROR: [17:58:07] ERROR: Could not sanitize molecule ending on line 258130
RDKit ERROR: [17:58:07] ERROR: Explicit valence for atom # 17 O, 3, is greater than permitted
RDKit ERROR: [17:58:07] Can't kekulize mol.  Unkekulized atoms: 57 58 59 60 61 62 63 64 65
RDKit ERROR: 
RDKit 

In [7]:
EGFR_ligands_ecfp = [Chem.GetMorganFingerprintAsBitVect(m, 2, nBits=1024)
                     for m in EGFR_ligands]

drugs_ecfp = [Chem.GetMorganFingerprintAsBitVect(m, 2, nBits=1024) for m in drugs]

In [8]:
EGFR_ligands_ecfp_sums = [0]*EGFR_ligands_ecfp[0].GetNumBits() # a list of zeros of a given length
for key in EGFR_ligands_ecfp:
    for onbit in key.GetOnBits():
        EGFR_ligands_ecfp_sums[onbit] += 1
EGFR_ligands_ecfp_sums

[54,
 1736,
 396,
 721,
 1561,
 109,
 63,
 345,
 1100,
 232,
 671,
 341,
 67,
 1755,
 1472,
 5298,
 103,
 25,
 86,
 319,
 217,
 335,
 96,
 216,
 101,
 889,
 412,
 236,
 82,
 506,
 51,
 1363,
 728,
 10397,
 1113,
 47,
 1353,
 74,
 76,
 357,
 167,
 933,
 1564,
 517,
 66,
 1265,
 510,
 68,
 83,
 99,
 201,
 287,
 61,
 583,
 152,
 405,
 392,
 223,
 571,
 190,
 188,
 1050,
 346,
 767,
 9072,
 1036,
 82,
 84,
 113,
 133,
 55,
 164,
 414,
 6266,
 811,
 1004,
 285,
 1335,
 178,
 271,
 8555,
 117,
 148,
 306,
 983,
 302,
 1042,
 425,
 73,
 211,
 2873,
 42,
 282,
 23,
 1323,
 81,
 510,
 294,
 218,
 74,
 104,
 75,
 825,
 63,
 367,
 94,
 209,
 981,
 44,
 144,
 349,
 121,
 1080,
 501,
 1288,
 365,
 719,
 812,
 285,
 1368,
 65,
 3729,
 247,
 220,
 79,
 248,
 208,
 106,
 11356,
 336,
 424,
 481,
 154,
 4101,
 249,
 294,
 2695,
 42,
 585,
 1902,
 948,
 73,
 256,
 223,
 321,
 428,
 125,
 2689,
 91,
 385,
 112,
 242,
 250,
 54,
 195,
 519,
 24,
 3602,
 485,
 88,
 275,
 1994,
 797,
 71,
 86,
 582,
 205,
 

In [9]:
drugs_ecfp_sums = [0]*drugs_ecfp[0].GetNumBits() # a list of zeros of a given length
for key in drugs_ecfp:
    for onbit in key.GetOnBits():
        drugs_ecfp_sums[onbit] += 1
drugs_ecfp_sums

[120,
 2360,
 285,
 317,
 939,
 231,
 29,
 138,
 213,
 303,
 260,
 258,
 68,
 665,
 227,
 828,
 103,
 104,
 93,
 187,
 118,
 203,
 63,
 237,
 154,
 162,
 127,
 111,
 62,
 211,
 69,
 172,
 159,
 4264,
 155,
 110,
 1076,
 73,
 87,
 207,
 79,
 440,
 648,
 112,
 79,
 261,
 319,
 76,
 52,
 97,
 89,
 123,
 245,
 271,
 100,
 63,
 105,
 85,
 158,
 178,
 69,
 179,
 204,
 219,
 2683,
 147,
 55,
 173,
 97,
 226,
 77,
 346,
 179,
 389,
 255,
 465,
 76,
 283,
 134,
 479,
 4939,
 170,
 152,
 122,
 313,
 87,
 270,
 124,
 108,
 105,
 1541,
 90,
 160,
 54,
 308,
 97,
 114,
 158,
 96,
 87,
 61,
 119,
 402,
 60,
 194,
 160,
 246,
 79,
 173,
 84,
 162,
 73,
 534,
 56,
 797,
 155,
 396,
 608,
 349,
 767,
 78,
 767,
 117,
 165,
 82,
 193,
 109,
 47,
 2503,
 89,
 337,
 123,
 98,
 162,
 127,
 52,
 625,
 56,
 515,
 399,
 667,
 59,
 125,
 150,
 289,
 199,
 99,
 2458,
 86,
 79,
 106,
 104,
 370,
 64,
 67,
 126,
 66,
 183,
 282,
 48,
 161,
 45,
 232,
 135,
 134,
 262,
 107,
 98,
 73,
 96,
 130,
 261,
 137,
 202,


In [10]:
EGFR_ligands_ecfp_scaled = [x/len(EGFR_ligands_ecfp) for x in EGFR_ligands_ecfp_sums]
drugs_ecfp_scaled = [x/len(drugs_ecfp) for x in drugs_ecfp_sums]

In [12]:
EGFR_drugbank_differences = [(i, a_b[0] - a_b[1])
                             for i, a_b in enumerate(zip(EGFR_ligands_ecfp_scaled, drugs_ecfp_scaled))]

EGFR_drugbank_differences.sort(key=lambda x: x[1])
EGFR_drugbank_differences[:3], EGFR_drugbank_differences[-3:]

([(1, -0.21081907504998498),
  (389, -0.16329466286013014),
  (147, -0.1582371317767479)],
 [(378, 0.519296000817336),
  (428, 0.5389652243666746),
  (491, 0.5792040720402662)])

The first three of the prevalent bits among EGFR ligands compared to DrugBank molecules are bits number 378, 428 and 491. The first three of the prevalent bits among DrugBank molecules compared to EGFR ligands are bits number 1, 389 and 147.

In [14]:
def substructures_of_bit_ecfp4_1024(mol, bit):
    bitinfo = {}
    Chem.GetMorganFingerprintAsBitVect(mol, 2, nBits=1024, bitInfo=bitinfo)
    hit_atoms = [Chem.FindAtomEnvironmentOfRadiusN(mol, radius, atom)
                 for atom, radius in bitinfo.get(bit, [])
                 if radius > 0]
    submols = [Chem.PathToSubmol(mol, ha) for ha in hit_atoms]
    return submols

In [15]:
substructure_smiles = {}
for ligand in EGFR_ligands:
    for submol in substructures_of_bit_ecfp4_1024(ligand, 378):
        subsmiles = Chem.MolToSmiles(submol)
        substructure_smiles[subsmiles] = substructure_smiles.get(subsmiles, 0) + 1
substructure_smiles

{'ccc(Cl)c(c)Cl': 260,
 'CCC(C)CO': 12,
 'ccn(c(c)c)C(C)C': 2,
 'C=C(C)C': 2,
 'C[C@@H](C)CC(C)(C)O': 1,
 'C=C(C)/C(=N/C)C(=C)O': 1,
 'ccc(cc)C(N)S': 1,
 'ccc(cc)[S+](c)[O-]': 1,
 'CN(C)[O]': 6,
 'cc(-c)cc(-c)n': 2,
 'C=CC(=O)c(c)c': 2}

In [16]:
substructure_smiles_db = {}
for drug in drugs:
    for submol in substructures_of_bit_ecfp4_1024(drug, 378):
        subsmiles = Chem.MolToSmiles(submol)
        substructure_smiles_db[subsmiles] = substructure_smiles_db.get(subsmiles, 0) + 1
substructure_smiles_db

{'C[C@@H](N)[C@@H](C)O': 2,
 'ccn(c(c)c)C(C)C': 5,
 'cc(c)[C@@H](O)CN': 2,
 'cc(c)[C@@H](Cc)[N+](C)(C)C': 4,
 'cc(c)C(Cc)[N+](C)(C)C': 2,
 'CCNC(c)C': 1,
 'ccc(Cl)c(c)Cl': 13,
 'COC(=O)CS': 1,
 'ccc(NS)c(c)-c': 1,
 'CC[NH3+]': 5,
 'ccc(sc)C(C)(C)N': 1,
 'C=C(S)C[C@H](C)N': 1,
 'C=CC[C@@H](C)O': 1,
 'CC[C@@H](NC)C(C)=O': 1,
 'CC(C)NC(C)c': 1,
 'C[C@@H](N)CCS': 2,
 'ccc(cc)C[NH3+]': 2,
 'CC(C)(C)SCN': 1,
 'ccn(nc)C(C)C': 1,
 'cc(c)[O-]': 6,
 'C=C(C)C': 4,
 'CC(=O)CC(=O)S': 1,
 'C[C@H](N)[C@@H](C)O': 2,
 '[NH3+]CO': 3,
 'ccc(C(=O)O)c(c)S': 1,
 'cc(o)Cl': 1,
 'cc(C)c(c(c)C)c(c)c': 2,
 'CCC(NC)C(C)=O': 2,
 'C[C@H](O)C[P@H](=O)[O-]': 1,
 'Cc(n)[nH]c([nH])=S': 1,
 'ccnc(n)N': 1,
 'ccc(cc)[C@@H](N)S': 1,
 'CN(C)C(=O)N(C)C': 1,
 'cn(c)O': 1,
 'CCNC(C)c': 1}

In [16]:
len(substructure_smiles_db)

35

In DrugBank and EGFR dataset the fragment present most often behind this bit is 'ccc(Cl)c(c)Cl'.